csrc/ir/nodes.cpp

```diff
@@ -2105,24 +2105,38 @@ std::vector<PolymorphicValue> MmaOp::evaluate(
       tv_b->getRootDomain().front()->isBroadcast(),
       "Expected first dimension to be broadcasted for second operand.");
 
+  // ATen preserves the dtype of MmaOp inputs whereas MmaOp generates float
+  // outputs. To preserve numerical equivalence and precision, the output of
+  // ATen matmul should be the same as MmaOp out `eventually`. Supported cases:
+  //  1. MmaOp->out() and MmaOp->input() are the same dtype.
+  //  2. MmaOp->out() is followed by a CastOp() to the MmaOp->input() dtype.
+  // NOTE: Currently MmaOp only accepts Half and BFloat16 so case (1) will not
+  // occur.
+
+  const std::vector<Expr*>& uses = out()->uses();
+  // Check if the only use of MmaOp is a single castOp.
+  bool is_use_single_cast = [&uses] {
+    return (
+        uses.size() == 1 && uses.front()->isA<UnaryOp>() &&
+        uses.front()->as<UnaryOp>()->getUnaryOpType() == UnaryOpType::Cast);
+  }();
+
+  // Check if we eventually convert to the ATen output dtype.
+  // See https://github.com/NVIDIA/Fuser/pull/1874#discussion_r1516991574
+  NVF_CHECK(
+      tv_a->getDataType() == out()->getDataType() ||
+      (is_use_single_cast &&
+       tv_a->getDataType() ==
+           uses.front()->as<UnaryOp>()->out()->getDataType()));
+
```

Comment:
```suggestion
  auto used_as_dtype = [](Val* v) -> DataType {
    const std::vector<Expr*>& uses = v->uses();
    if (uses.size() == 1) {
      if (auto* unary = dynamic_cast<UnaryOp*>(uses.front())) {
        if (unary->getUnaryOpType() == UnaryOpType::Cast) {
          return unary->out()->getDataType();
        }
      }
    }
    return v->getDataType();
  };
  NVF_CHECK(used_as_dtype(out()) == tv_a->getDataType());
```

(The above is just to show the idea -- I didn't even try to compile it)